README.md
<p align="center">
<img src="https://assets.solidjs.com/banner?project=icons" height="150">
</p>

<h1 align="center">
Solid Icons
</h1>
<p align="center">
Modern solution for use icons on SolidJS
<p>

<p align="center">
  <a href="https://www.npmjs.com/package/solid-icons?activeTab=versions"><img src="https://badgen.net/npm/v/solid-icons"></a>
  <a href="https://github.com/x64Bits/solid-icons/blob/main/LICENSE"><img src="https://badgen.net/npm/license/solid-icons"></a>
  <a href="https://www.npmjs.com/package/solid-icons"><img src="https://badgen.net/npm/dt/solid-icons"></a>
<p>

<p align="center">
 <a href="https://solid-icons.vercel.app/">Icons Explorer</a> | <a href="https://github.com/x64Bits/solid-icons-web">Example</a> 
</p>

<h4 align="center">

</h4>
<br>
<br>

## ✨ Features

- [16 Icon packs](#-included-icons-pack) totally ready to use.
- Compatible with [Solid Start](https://github.com/solidjs/solid-start) static generation and SSR.
- Tree shakeable: What you take is what you get.
- [Customizable](#%EF%B8%8F-configuration) - receive props to extend their usefulness.
- [Reactivity](https://www.youtube.com/watch?v=J70HXl1KhWE), take advantage of SolidJS to react to changes in props.
- [Just import and declare](#usage) in your JSX to work out-the-box
- First class TypeScript support

## 📦 Installation

### Yarn

```bash
yarn add solid-icons
```

### NPM

```bash
npm install solid-icons --save
```

## Usage

```jsx
import { TbBrandSolidjs } from "solid-icons/tb";

<TbBrandSolidjs size={24} color="#2c4f7c" />;
```

## Custom icon

There are situations where you want to use your own set of icons, CustomIcon is a component exposed from the library that uses the IconTemplate that all the icons in the library already use. (Thanks [kdaquila](https://github.com/kdaquila) for the example).

```jsx
import { CustomIcon } from "solid-icons/lib";

const iconContent = {
  a: { fill: "currentColor", viewBox: "0 0 384 512" },
  c: '<path d="M384 319.1C384 425.9 297.9 512 192 512S0 425.87 0 320c0-58.67 27.82-106.8 54.57-134.1C69.54 169.3 96 179.8 96 201.5V287c0 35.17 27.97 64.5 63.16 64.94C194.9 352.5 224 323.6 224 288c0-88-175.1-96.12-52.15-277.2C185.35-8.92 216 .03 216 23.83 215.1 127 384 149.7 384 319.1z"/>',
}

<CustomIcon src={iconContent} size={24} color="#2c4f7c" />;
```

### Props

| Key     | Default     | Notes                                                                                            |
| ------- | ----------- | ------------------------------------------------------------------------------------------------ |
| `src`   | required    | format: `a` Attributes needed for the svg like `viewBox`, `c`: svg content, look at the example. |
| `size`  | `1em`       |                                                                                                  |
| `class` | `undefined` |                                                                                                  |
| `title` | `undefined` | A icon title a11y                                                                                |

## 🔋 Included icons pack

| Icon Library                                                       | License                                                                 | Version | Abbreviation |
| ------------------------------------------------------------------ | ----------------------------------------------------------------------- | ------- | ------------ |
| [Ant Design Icons](https://github.com/ant-design/ant-design-icons) | [MIT](https://opensource.org/licenses/MIT)                              | 4.2.1   | ai           |
| [Bootstrap Icons](https://github.com/twbs/icons)                   | [MIT](https://opensource.org/licenses/MIT)                              | 1.10.5  | bs           |
| [BoxIcons](https://github.com/atisawd/boxicons)                    | [CC BY 4.0 License](https://creativecommons.org/licenses/by/4.0/)       | 2.1.4   | bi           |
| [Feather](https://feathericons.com/)                               | [MIT](https://github.com/feathericons/feather/blob/master/LICENSE)      | 4.29.0  | fi           |
| [Font Awesome](https://fontawesome.com/)                           | [CC BY 4.0 License](https://creativecommons.org/licenses/by/4.0/)       | 6.4.0   | fa           |
| [Heroicons](https://github.com/refactoringui/heroicons)            | [MIT](https://github.com/tailwindlabs/heroicons/blob/master/LICENSE)    | 2.0.18  | hi           |
| [IcoMoon Free](https://github.com/Keyamoon/IcoMoon-Free)           | [CC BY 4.0 License](https://creativecommons.org/licenses/by/4.0/)       | 1.0.0   | im           |
| [Ionicons](https://ionicons.com/)                                  | [MIT](https://github.com/ionic-team/ionicons/blob/master/LICENSE)       | 7.1.2   | io           |
| [Remix Icon](https://github.com/Remix-Design/RemixIcon)            | [Apache License Version 2.0](http://www.apache.org/licenses/)           | 3.3.0   | ri           |
| [Simple Icons](https://simpleicons.org/)                           | [CC0 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/) | 9.0.0   | si           |
| [Typicons](http://s-ings.com/typicons/)                            | [CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/)         | 2.1.2   | ti           |
| [VS Code Icons](https://github.com/microsoft/vscode-codicons)      | [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)               | 0.0.32  | vs           |
| [Weather Icons](https://erikflowers.github.io/weather-icons/)      | [SIL OFL 1.1](http://scripts.sil.org/OFL)                               | 2.0.12  | wi           |
| [css.gg](https://github.com/astrit/css.gg)                         | [MIT](https://opensource.org/licenses/MIT)                              | 2.0.0   | cg           |
| [Tabler Icons](https://github.com/tabler/tabler-icons)             | [MIT](https://opensource.org/licenses/MIT)                              | 2.20.0  | tb           |
| [Github Octicons](https://github.com/primer/octicons)              | [MIT](https://opensource.org/licenses/MIT)                              | 19.1.0  | oc           |

## ⚙️ Configuration

solid-icons components receive props like any SVG, you also have a few custom ones.

```jsx
import { TbBrandSolidjs } from "solid-icons/tb";

<TbBrandSolidjs size={24} color="#2c4f7c" class="custom-icon" title="a11y" />;
```

| Key     | Default                  | Notes             |
| ------- | ------------------------ | ----------------- |
| `color` | `currentColor` (inherit) |                   |
| `size`  | `1em`                    |                   |
| `class` | `undefined`              |                   |
| `title` | `undefined`              | A icon title a11y |

## 💻 Development

requirements:

node **^16.14.0**

### Basic build

You can locally clone this repository:

```bash
$ git clone https://github.com/x64Bits/solid-icons
$ cd solid-icons
$ yarn
$ yarn build
```

### Build dev mode

If you did the above steps and want to build while listening if the files change you can run:

```bash
$ yarn dev
```

### Supported arguments

Isolate a single library, this allows you to avoid recompiling the entire library and thus optimize the result of a single one:

```bash
$ yarn dev --isolate="ai"
```

This command is used to build the files destined for web, if you want to change the path, in `src/build/constants.ts` you can modify the output of the files:

```bash
$ yarn dev --web
```

## 📝 Licence

MIT

- Icons are taken from the other projects so please check each project licences accordingly.


CHANGELOG
## CHANGELOG 1.0.9:

- linkthai introduce a solution for latest tabler issue on `stroke-width=2` by default.

## CHANGELOG 1.0.8:

- Fix remix color doesn't affect by `color` prop.

## CHANGELOG 1.0.5:

- Fix csss color selector.
- Add CustomIcon component.

## CHANGELOG 1.0.0:

- Replace build legacy build strategy.
- Fix Octicons naming convention.
- Standardize dist files with the ecosystem.
- tabler only replace one stroke.

## CHANGELOG 0.4.2:

- Fixed a bug where the reactivity was not working when passing props.

## CHANGELOG 0.4.1:

- Fixed most of the ionicons icons that had problems.

- The following icon libraries have been updated:
  - ionicons

This update adds 14 new icons

## CHANGELOG 0.3.3:

- The following icon libraries have been updated:
  - ant-design-icons
  - boxicons
  - feather
  - font-awesome
  - heroicons
  - bootstrap-icons
  - simple-icons
  - vscode-codicons
  - weather-icons

This update adds 398 new icons

## CHANGELOG 0.3.2:

- Improved import with autocomplete

## CHANGELOG 0.3.1:

- Accept prop for external viewBox
- Remove unused types

## CHANGELOG 0.3.0:

- Fix Heroicons and Ant Design fill/stroke
- [Icons showcase website](https://solid-icons-web.vercel.app/) is live 🚀
- IconWrapper main component accept reactive props
- Most of the icon packs have been updated to the latest version.


scripts/build.js
var cheerio = require("cheerio");
if (typeof cheerio != "function") cheerio = require("cheerio").default;
const path = require("path");
const fs = require("fs");
const fsPromise = require("fs").promises;
const { promisify } = require("util");
const exec = promisify(require("child_process").exec);
const chalk = require("chalk");

const mkdir = promisify(fs.mkdir);
const readdir = promisify(fs.readdir);
const copyFile = promisify(fs.copyFile);
const appendFile = promisify(fs.appendFile);

let totalCount = 0;

const {
  templateType,
  WEB_DIR,
  PACKAGE_JSON_PACK,
  OUT_DIR,
} = require("./constants");

// Utils
const {
  generateSearchFile,
  makeIconFile,
  generateMetaPackFile,
} = require("./files");

const viewBoxes = {
  bi: "0 0 24 24",
  vsc: "0 0 16 16",
};

const _rootDir = path.resolve(__dirname, "../");

const projectPath = path.resolve(__dirname, "../");

// LICENSE
const licenseFrom = path.resolve(projectPath, "LICENSE");
const licenseTo = path.resolve(OUT_DIR, "LICENSE");

// README.md
const readmeFrom = path.resolve(projectPath, "README.md");
const readmeTo = path.resolve(OUT_DIR, "README.md");

// Manifest
const iconManifest = require("../manifest.json");

const { normalizeIcon, normalizeTbIcon } = require("./normalize");
const { moveLibraryArtifacts } = require("./lib");
const manifestInfo = {};
const log = console.log;

function capitalizeFirstLetter(_string) {
  const fixName = {
    filled: "fill",
    outlined: "outline",
    regular: "",
    logos: "logo",
  };
  const string = fixName[_string] !== undefined ? fixName[_string] : _string;
  return string.charAt(0).toUpperCase() + string.slice(1);
}

async function loadSvgFilesList({ svgPath, prefix }, iconPack) {
  const list = await readdir(svgPath);
  const outList = [];

  for (const f of list) {
    const [_fileName, fileExt] = f.split(".");
    if (fileExt && fileExt.toLowerCase() === "svg") {
      let fileName = _fileName;
      if (
        Array.isArray(iconPack.removeFilePrefixes) &&
        iconPack.removeFilePrefixes.length > 0
      ) {
        for (const pfix of iconPack.removeFilePrefixes) {
          if (pfix && fileName.indexOf(pfix) === 0) {
            fileName = fileName.substring(pfix.length);
          }
        }
      }

      if (iconPack.removeFirtsCharsFromFile > 0) {
        fileName = fileName.substring(iconPack.removeFirtsCharsFromFile);
      }

      outList.push({
        filePath: path.resolve(svgPath, f),
        svgName:
          prefix +
          fileName
            .split("-")
            .map((x) => capitalizeFirstLetter(x))
            .join(""),
      });
    }
  }

  return outList;
}

function filterFolders(dir) {
  if ([".DS_Store"].includes(dir)) return false;
  return true;
}

function filterAttr(attr) {
  if (["p-id", "t", "style", "id"].includes(attr)) return false;
  return true;
}

function replaceColor(attr, sn) {
  const BLACK_COLOR = "#000";
  if (attr.includes(BLACK_COLOR) && sn === "io") {
    return attr.replace(BLACK_COLOR, "currentColor");
  }

  if (attr.includes(`fill="#`)) {
    if (sn === "io") console.log(attr);
    return ` fill="currentColor"`;
  }
  return attr;
}

function tagTreeToString(tagData, sn) {
  if (["defs"].includes(tagData.tag)) return "";

  return `<${tagData.tag}${Object.keys(tagData.attr)
    .filter(filterAttr)
    .map((k) => ` ${k}="${tagData.attr[k]}"`)
    .map((r) => replaceColor(r, sn))
    .join("")}>${(tagData.child || [])
    .map((t) => tagTreeToString(t))
    .filter((x) => !!x)
    .join("\n")}</${tagData.tag}>`;
}

async function generateSvgIconInfo(
  fileName,
  iconData,
  shortName,
  exportType = "lib"
) {
  if (!iconData.attr.viewBox) {
    iconData.attr.viewBox = viewBoxes[shortName];
  }

  const comressed = {
    a: {
      stroke: "none",
      ...iconData.attr,
    },
    c: (iconData.child || [])
      .map((t) => tagTreeToString(t, shortName))
      .map((x) => {
        if (shortName === "io" && x.includes("<line")) {
          const lineFixed = x.replace("<line", '<line stroke="currentColor"');

          return lineFixed;
        }

        return x;
      })
      .filter((x) => !!x)
      .join(""),
  };

  const templateExport = templateType[exportType];

  switch (shortName) {
    case "io": {
      const normalizedIo = await normalizeIcon(comressed, fileName);

      return templateExport(fileName, normalizedIo);
    }
    case "tb": {
      const normalizedTb = normalizeTbIcon(comressed);

      return templateExport(fileName, normalizedTb);
    }

    default:
      return templateExport(fileName, comressed);
  }
}

async function convertIconData(svg, multiColor) {
  const $svg = cheerio.load(svg, { xmlMode: true })("svg");

  // filter/convert attributes
  // 1. remove class attr
  const attrConverter = (
    /** @type {{[key: string]: string}} */ attribs,
    /** @type string */ tagName
  ) =>
    attribs &&
    Object.keys(attribs)
      .filter(
        (name) =>
          ![
            "class",
            ...(tagName === "svg"
              ? ["xmlns", "xmlns:xlink", "xml:space", "width", "height"]
              : []), // if tagName is svg remove size attributes
          ].includes(name)
      )
      .reduce((obj, name) => {
        const newName = name;
        switch (newName) {
          case "fill":
            if (
              attribs[name] === "none" ||
              attribs[name] === "currentColor" ||
              multiColor
            ) {
              obj[newName] = attribs[name];
            }
            break;
          default:
            obj[newName] = attribs[name];
            break;
        }
        return obj;
      }, {});

  // convert to [ { tag: 'path', attr: { d: 'M436 160c6.6 ...', ... }, child: { ... } } ]
  const elementToTree = (/** @type {Cheerio} */ element) =>
    element
      .filter((_, e) => e.tagName && !["style"].includes(e.tagName))
      .map((_, e) => {
        if (e.tagName === "path") {
          if (e.attribs.stroke) {
            e.attribs.stroke = "currentColor";
          }
        }
        return {
          tag: e.tagName,
          attr: attrConverter(e.attribs, e.tagName),
          child:
            e.children && e.children.length
              ? elementToTree(cheerio(e.children))
              : undefined,
        };
      })
      .get();

  const tree = elementToTree($svg);
  return tree[0];
}

async function generateSvg(iconPack, svgItem) {
  const svgStr = await promisify(fs.readFile)(svgItem.filePath, "utf8");
  const iconData = await convertIconData(svgStr, true);
  return iconData;
}

async function loadPack(iconPack) {
  manifestInfo[iconPack.shortName] = {
    iconsList: [],
    name: iconPack.packName,
    path: iconPack.shortName,
    license: iconPack.license,
    sourceUrl: iconPack.url,
  };

  const packFolder = path.resolve(`${OUT_DIR}/${iconPack.shortName}`);
  await mkdir(packFolder);

  const packPath = path.resolve(`${OUT_DIR}/${iconPack.shortName}`);

  // Write webside artifacts
  const webPackFolder = path.resolve(`${WEB_DIR}/icons`, iconPack.shortName);
  await mkdir(webPackFolder);

  // package.json pack

  appendFile(path.resolve(packPath, "package.json"), PACKAGE_JSON_PACK);

  // Icon File
  const headerFile = `import { IconTemplate } from "../lib";`;
  appendFile(path.resolve(packPath, `index.module.js`), headerFile);

  // Icon CJS file
  const headerCjsFile = `var IconTemplate = require('../lib').IconTemplate;`;
  appendFile(path.resolve(packPath, `index.js`), headerCjsFile);

  // TS File
  const headerTsFile = `import type { IconTypes } from '../lib/browser/IconWrapper'`;
  appendFile(path.resolve(packPath, `index.d.ts`), headerTsFile);

  const baseFolder = path.resolve(__dirname, "../", iconPack.iconsPath);

  const folders = [];
  if (iconPack.subFolders) {
    // load subfolders
    const list = await readdir(baseFolder);
    for (const li of list.filter(filterFolders)) {
      folders.push({
        prefix:
          capitalizeFirstLetter(
            iconPack.removeMainPrefix ? "" : iconPack.shortName
          ) + capitalizeFirstLetter(li.toLowerCase()),
        svgPath: path.resolve(baseFolder, li),
      });
    }
  } else {
    folders.push({
      prefix: capitalizeFirstLetter(
        iconPack.removeMainPrefix ? "" : iconPack.shortName
      ),
      svgPath: baseFolder,
    });
  }

  log(
    chalk.dim(`${iconPack.packName} package has been generated`) +
      chalk.green(" ✓")
  );

  for (const item of folders) {
    const svgList = await loadSvgFilesList(item, iconPack);

    for (const svgFile of svgList) {
      manifestInfo[iconPack.shortName].iconsList.push(svgFile.svgName);

      const iconData = await generateSvg(iconPack, svgFile);

      const svgAsJs = await generateSvgIconInfo(
        svgFile.svgName,
        iconData,
        iconPack.shortName
      );

      const svgAsCjs = await generateSvgIconInfo(
        svgFile.svgName,
        iconData,
        iconPack.shortName,
        "cjs"
      );

      const defaultSvg = await generateSvgIconInfo(
        svgFile.svgName,
        iconData,
        iconPack.shortName,
        "isolate"
      );

      await makeIconFile(defaultSvg, iconPack.shortName, svgFile.svgName);

      // Icon File ESM
      appendFile(path.resolve(packPath, `index.module.js`), svgAsJs);

      // Icon File Server side
      appendFile(path.resolve(packPath, `index.js`), svgAsCjs);

      // TS File
      const contentTsFile = `\nexport declare const ${svgFile.svgName}: IconTypes;`;
      appendFile(path.resolve(packPath, `index.d.ts`), contentTsFile);

      totalCount++;
    }
  }
}

async function buildLib() {
  const execOpt = {
    cwd: _rootDir,
  };

  await Promise.all([
    exec("rollup -c", execOpt),
    exec(`rm -rf ${WEB_DIR}/icons && mkdir ${WEB_DIR}/icons`),
    exec(`rm -rf ${WEB_DIR}/search.js && touch ${WEB_DIR}/search.js`),
    exec(`rm -rf ${WEB_DIR}/meta.js && touch ${WEB_DIR}/meta.js`),
  ]);

  await moveLibraryArtifacts();
}

async function writePackageJson() {
  const DIST = path.resolve(_rootDir, "dist");

  const packageJsonStr = await fsPromise.readFile(
    path.resolve(_rootDir, "package.json"),
    "utf-8"
  );
  let packageJson = JSON.parse(packageJsonStr);

  delete packageJson.private;
  delete packageJson.devDependencies;
  delete packageJson.scripts;

  packageJson = {
    ...packageJson,
  };

  const editedPackageJsonStr = JSON.stringify(packageJson, null, 2);
  await fsPromise.writeFile(
    path.resolve(DIST, "package.json"),
    editedPackageJsonStr
  );
}

async function generateIconsManifest() {
  await appendFile(
    path.resolve(WEB_DIR, `meta.js`),
    `export default [\n`,
    "utf-8"
  );
  for (const pk of Object.keys(manifestInfo)) {
    const pack = manifestInfo[pk];

    for (const prop of Object.keys(pack)) {
      if (prop === "iconsList") {
        await generateSearchFile(pack);

        await generateMetaPackFile(pack);
      }
    }
  }
  await fsPromise
    .readFile(path.resolve(WEB_DIR, `search.js`), "utf-8")
    .then(async (file) => {
      await fsPromise.writeFile(
        path.resolve(WEB_DIR, `search.js`),
        `export default ${file}`,
        "utf-8"
      );
    });
  await appendFile(path.resolve(WEB_DIR, `meta.js`), `]`, "utf-8");
}

async function init() {
  await buildLib();

  await writePackageJson();

  await copyFile(licenseFrom, licenseTo);

  await copyFile(readmeFrom, readmeTo);

  console.log("Artifacts files has been copied");
}

async function main() {
  console.log("Initializing");
  await init();

  for (const iconPack of iconManifest.sort((x1, x2) => {
    if (x1.packName > x2.packName) return 1;
    if (x1.packName < x2.packName) return -1;
    return 0;
  })) {
    await loadPack(iconPack);
  }

  await generateIconsManifest();

  log(" ");

  log(chalk.green(`${totalCount}`) + chalk.dim(" icons has been generated"));

  console.log(`  `);
}
main();
